src/testbase.rs

```diff
@@ -0,0 +1,300 @@
+//! Common utilities shared by unit tests.
+#![cfg(test)]
+
+use std::collections::HashSet;
+
+use nalgebra::{Point3, Vector3};
+
+use aabb::{AABB, Bounded};
+use bounding_hierarchy::BoundingHierarchy;
+use ray::Ray;
+
+/// A vector represented as a tuple
+pub type TupleVec = (f32, f32, f32);
+
+/// Convert a `TupleVec` to a `nalgebra` point.
+pub fn tuple_to_point(tpl: &TupleVec) -> Point3<f32> {
+    Point3::new(tpl.0, tpl.1, tpl.2)
+}
+
+/// Convert a `TupleVec` to a `nalgebra` vector.
+pub fn tuple_to_vector(tpl: &TupleVec) -> Vector3<f32> {
+    Vector3::new(tpl.0, tpl.1, tpl.2)
+}
+
+/// Define some `Bounded` structure.
+pub struct UnitBox {
+    pub id: i32,
+    pub pos: Point3<f32>,
+}
+
+/// `UnitBox`'s `AABB`s are unit `AABB`s centered on the given x-position.
+impl Bounded for UnitBox {
+    fn aabb(&self) -> AABB {
+        let min = self.pos + Vector3::new(-0.5, -0.5, -0.5);
+        let max = self.pos + Vector3::new(0.5, 0.5, 0.5);
+        AABB::with_bounds(min, max)
+    }
+}
+
+/// Generate 21 `UnitBox`s along the X axis centered on whole numbers (-10,9,..,10).
+/// The index is set to the rounded x-coordinate of the box center.
+pub fn generate_aligned_boxes() -> Vec<UnitBox> {
+    // Create 21 boxes along the x-axis
+    let mut shapes = Vec::new();
+    for x in -10..11 {
+        shapes.push(UnitBox {
+            id: x,
+            pos: Point3::new(x as f32, 0.0, 0.0),
+        });
+    }
+    shapes
+}
+
+/// Creates a `BoundingHierarchy` for a fixed scene structure.
+pub fn build_some_bh<BH: BoundingHierarchy>() -> (Vec<UnitBox>, BH) {
+    let boxes = generate_aligned_boxes();
+    let bh = BH::build(&boxes);
+    (boxes, bh)
+}
+
+/// Given a ray, a bounding hierarchy, the complete list of shapes in the scene and a list of
+/// expected hits, verifies, whether the ray hits only the expected shapes.
+fn traverse_and_verify<BH: BoundingHierarchy>(ray_origin: Point3<f32>,
+                                              ray_direction: Vector3<f32>,
+                                              all_shapes: &Vec<UnitBox>,
+                                              bh: &BH,
+                                              expected_shapes: &HashSet<i32>) {
+    let ray = Ray::new(ray_origin, ray_direction);
```

Comment:
It does not make any semantic difference, but with this interface we can reduce a little more duplicate code (which is the only purpose of exporting the common functionality as a function).